### PR TITLE
replace horovod by estimator

### PR DESCRIPTION
**Pull request by fsx950223**

### CODE REVIEW

File: efficientdet/det_model_fn.py [hunk 1]
```diff
@@ -590,21 +585,22 @@ def scaffold_fn():
...
-    profile_hook = tf.train.ProfilerHook(
-        save_steps=1000, output_dir=params['model_dir'])
-    training_hooks.append(profile_hook)
+    if params.get('profile', False):
+      profile_hook = tf.estimator.ProfilerHook(
+          save_steps=1000, output_dir=params['model_dir'], show_memory=True)
+      training_hooks.append(profile_hook)
 
-    # Report memory allocation if OOM
-    class OomReportingHook(tf.estimator.SessionRunHook):
```
Comment (kartik4949) [inline efficientdet/det_model_fn.py:598]: if profile is false , it seems we wont have OOM hook

File: efficientdet/dataloader.py [hunk 2]
```diff
@@ -333,12 +333,8 @@ def dataset_parser(self, value, example_decoder, anchor_labeler, params):
       image_scale = input_processor.image_scale_to_original
       boxes *= image_scale
       is_crowds = tf.cast(is_crowds, dtype=tf.float32)
-      boxes = pad_to_fixed_size(boxes, -1, [self._max_instances_per_image, 4])
```
Comment (mingxingtan) [inline efficientdet/dataloader.py:336]: What are these changes?

File: efficientdet/dataloader.py [hunk 3]
```diff
@@ -420,7 +421,40 @@ def _prefetch_dataset(filename):
...
-    dataset = dataset.batch(batch_size, drop_remainder=True)
+    shapes = list(tf.nest.map_structure(lambda item: item.shape,
+                                        dataset.element_spec))
+    shapes[6] = (self._max_instances_per_image,4)
+    shapes[7] = (self._max_instances_per_image,1)
+    shapes[8] = (self._max_instances_per_image,1)
+    shapes[9] = (self._max_instances_per_image,1)
+    shapes = tuple(shapes)
+
+    def make_zero(t):
```
Comment (mingxingtan) [inline efficientdet/dataloader.py:432]: These codes look complicated, could you help explain a little bit?

File: efficientdet/main.py [hunk 4]
```diff
@@ -240,19 +236,27 @@ def _can_partition(spatial_dim):
...
-  run_config = tf.estimator.tpu.RunConfig(
-      cluster=tpu_cluster_resolver,
-      model_dir=model_dir,
-      log_step_count_steps=FLAGS.iterations_per_loop,
-      session_config=config_proto,
-      tpu_config=tpu_config,
-      tf_random_seed=FLAGS.tf_random_seed,
-  )
+    strategy = \
+      tf.distribute.MirroredStrategy() if FLAGS.strategy == 'gpus' else None
```
Comment (mingxingtan) [inline efficientdet/main.py:251]: Try avoid "\", maybe something like:

strategy = None
if FLAGS.strategy == 'gpus':
   strategy = tf.distribute.MirroredStrategy()

**Review by mingxingtan — COMMENTED** [review]
@fsx950223 
Nice! Does it work on multi-GPUs?

File: efficientdet/dataloader.py (hunk 3 above)
Comment (fsx950223) [reply efficientdet/dataloader.py:432]: I used batch_padded replace pad_to_fixed_size which faster than later.

**Comment by fsx950223** [comment]
> @fsx950223
> Nice! Does it work on multi-GPUs?

I have train the model with voc dataset on 2x 1080ti.

File: efficientdet/utils.py [hunk 5]
```diff
@@ -296,8 +275,9 @@ def call(self, inputs, training=None):
 def batch_norm_class(is_training, strategy=None):
   if is_training and strategy == 'tpu':
     return TpuBatchNormalization
-  elif is_training and strategy == 'horovod':
-    return SyncBatchNormalization
+  elif is_training and strategy == 'gpus':
+    # TODO(fsx950223): Wait for #41980 to be fixed
```
Comment (mingxingtan) [inline efficientdet/utils.py:279]: What does this mean?

File: efficientdet/main.py [hunk 6]
```diff
@@ -284,13 +297,21 @@ def _eval(steps):
...
+    if FLAGS.strategy == 'tpu':
+      eval_estimator = tf.estimator.tpu.TPUEstimator(
+          model_fn=model_fn_instance,
+          use_tpu=use_tpu,
+          train_batch_size=FLAGS.train_batch_size,
+          eval_batch_size=FLAGS.eval_batch_size,
+          config=run_config,
+          params=eval_params)
+    else:
+      eval_params['batch_size'] = \
```
Comment (mingxingtan) [inline efficientdet/main.py:309]: Prefer something like

eval_params['batch_size'] =  (
    FLAGS.eval_batch_size // getattr(strategy, 'num_replicas_in_sync', 1))

Internally, we almost never use "\" (It is a warning)

File: efficientdet/dataloader.py (hunk 3 above)
Comment (mingxingtan) [reply efficientdet/dataloader.py:432]: Aha, gotcha. Could you add a simple comment like "Shape and paddings are used for padded_batch."

**Review by mingxingtan — APPROVED** [review]
Overall looks good.

File: efficientdet/utils.py (hunk 5 above)
Comment (fsx950223) [reply efficientdet/utils.py:279]: Tensorflow issue #41980.
It's a bug in keras.syncbatchnormalization which is caused by incorrect implementation of nccl all_reduce

File: efficientdet/dataloader.py [hunk 7]
```diff
@@ -333,12 +308,8 @@ def dataset_parser(self, value, example_decoder, anchor_labeler, params):
...
       boxes *= image_scale
       is_crowds = tf.cast(is_crowds, dtype=tf.float32)
-      boxes = pad_to_fixed_size(boxes, -1, [self._max_instances_per_image, 4])
-      is_crowds = pad_to_fixed_size(is_crowds, 0,
-                                    [self._max_instances_per_image, 1])
-      areas = pad_to_fixed_size(areas, -1, [self._max_instances_per_image, 1])
-      classes = pad_to_fixed_size(classes, -1,
-                                  [self._max_instances_per_image, 1])
+      is_crowds = tf.expand_dims(is_crowds, axis=1)
+      areas = tf.expand_dims(areas, axis=1)
```
Comment (mingxingtan) [inline efficientdet/dataloader.py:312]: How faster is this new dataloader way?
I feel the code is much harder to read than before. If it is not significantly faster, maybe we can stick to the simple way for better readability?

File: efficientdet/utils.py (hunk 5 above)
Comment (mingxingtan) [reply efficientdet/utils.py:279]: Could you replace it with something like:

```
# TODO(fsx950223): use SyncBatchNorm after the TF bug (incorrect nccl all_reduce) is fixed:
#  see https://github.com/tensorflow/tensorflow/issues/41980
```

File: efficientdet/main.py [hunk 8]
```diff
@@ -117,17 +121,8 @@
...
-    import horovod.tensorflow as hvd  # pylint: disable=g-import-not-at-top
-    logging.info('Use horovod with multi gpus')
-    hvd.init()
-    os.environ['CUDA_VISIBLE_DEVICES'] = str(hvd.local_rank())
-  import tensorflow.compat.v1 as tf  # pylint: disable=g-import-not-at-top
-  tf.enable_v2_tensorshape()
-  tf.disable_eager_execution()
-
   if FLAGS.strategy == 'tpu':
+    tf.disable_eager_execution()
```
Comment (fsx950223) [inline efficientdet/main.py:125]: I could train the model with global eager execution enabled.
Does ```TpuEstimator``` support it?

**Review by mingxingtan — APPROVED** [review]
Looks great!

**Comment by ice-tong** [comment]
Hi @fsx950223, I'm late to this game. I wonder why horovod is deprecated. For performance issues? 🤣

**Comment by fsx950223** [comment]
Yeah, bad performance when the model work with horovod sync batchnorm.
